src/connector/src/sink/snowflake.rs

```diff
@@ -236,31 +227,87 @@ impl SnowflakeSinkWriter {
+            generate_s3_file_name(self.s3_client.s3_path(), self.file_suffix.as_ref().unwrap());
+        let uploader = self
+            .s3_client
+            .opendal_s3_engine
+            .streaming_upload(&path)
+            .await
+            .map_err(|err| {
+                SinkError::Snowflake(format!(
+                    "failed to create the streaming uploader of opendal s3 engine for epoch {}, error: {}",
+                    self.epoch,
+                    err
+                ))
+            })?;
+        self.streaming_uploader = Some(uploader);
+        // we don't have data at the beginning of each epoch
+        self.has_data = false;
+        Ok(())
     }
 
-    fn at_sink_threshold(&self) -> bool {
-        self.row_counter >= self.max_batch_row_num
+    /// write data to the current streaming uploader for this epoch.
+    async fn streaming_upload(&mut self, data: Bytes) -> Result<()> {
+        let Some(uploader) = self.streaming_uploader.as_mut() else {
+            return Err(SinkError::Snowflake(format!(
+                "expect streaming uploader to be properly initialized when performing streaming upload for epoch {}",
+                self.epoch
+                ))
+            );
+        };
+        uploader
+            .write_bytes(data)
+            .await
+            .map_err(|err| {
+                SinkError::Snowflake(format!(
+                    "failed to write bytes when streaming uploading to s3 for snowflake sink, error: {}",
+                    err
+                ))
+            })?;
+        // well, at least there are some data to be uploaded
+        self.has_data = true;
+        Ok(())
     }
 
-    fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+    /// finalize streaming upload for this epoch.
+    /// ensure all the data has been properly uploaded to intermediate s3.
+    async fn finish_streaming_upload(&mut self) -> Result<()> {
+        let uploader = std::mem::take(&mut self.streaming_uploader);
+        let Some(uploader) = uploader else {
+            return Err(SinkError::Snowflake(format!(
+                "streaming uploader is not valid when trying to finish streaming upload for epoch {}",
+                self.epoch
+                )
+            ));
+        };
+        uploader.finish().await.map_err(|err| {
+            SinkError::Snowflake(format!(
+                "failed to finish streaming upload to s3 for snowflake sink, error: {}",
+                err
+            ))
+        })?;
+        Ok(())
+    }
+
+    async fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
         for (op, row) in chunk.rows() {
             assert_eq!(op, Op::Insert, "expect all `op(s)` to be `Op::Insert`");
             let row_json_string = Value::Object(self.row_encoder.encode(row)?).to_string();
-            self.payload.push_str(&row_json_string);
-            self.row_counter += 1;
+            self.streaming_upload(row_json_string.into()).await?;
```

Comment:
Better to use a `BytesMut` and write to `BytesMut` first, and then at the end call `streaming_upload` for only once so that we can reduce the number of bytes allocations. It will be better if `Value::Object(...).to_string()` can be replaced with extending to the `BytesMut` buffer directly so that the unnecessary string allocation can also be avoided.

Reply:
then what's the difference between this approach and the original bulk load approach?
if we are just uploading data to `streaming_uploader` in a all-at-once manner rather than on-the-fly, then probably there is no difference whether we use streaming upload or not?

refer: https://github.com/risingwavelabs/risingwave/pull/16269#discussion_r1565270840.

Reply:
This is not all-at-once at all. 

There are multiple stream chunks between two checkpoint barriers. The whole set of data to upload is the data between two checkpoint barriers rather than each individual stream chunk. This comment is just to request change on the granularity of on-the-fly streaming upload from row-by-row to chunk-by-chunk. The benefit is to reduce bytes allocation mentioned in the original comment, and also the allocation of box future (async-trait macro on the streaming uploader trait will turn the future of each async call to a box future). Doing allocation row-by-row may affect the performance.

Plus: the benefit gained from replacing bulk-load:
* No need to maintain the buffer by our own, cleaner implementation.
* Less files. Let's say we have 1M rows to upload, and buffer is 10k rows. If we do the original bulk upload, we will generate 100 small files. With streaming upload, there is only one file.

Reply:
> It will be better if Value::Object(...).to_string() can be replaced with extending to the BytesMut buffer directly so that the unnecessary string allocation can also be avoided.

this seems semantically in-consistent with snowflake side 🤔, i.e.,

1. since we need the string representation of the json row to let `snowflake pipe` correctly identify and ingest the corresponding data from intermediate s3 bucket - but a bunch of random bytes that can't be correctly interpreted (e.g., a serialized version of `serde_json::Map`) does not seem to be correct.

an example is as below - this is the content for a single intermediate file in s3 bucket.
![CleanShot 2024-04-17 at 14 28 08@2x](https://github.com/risingwavelabs/risingwave/assets/91381303/98be9abb-3d61-4349-86c2-3d0efd6d4034)

2. starrocks sink is using the same approach to serialize json object along the way, i.e., https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/sink/starrocks.rs#L335

Reply:
The implementation of `ToString` is [inherited](https://docs.rs/serde_json/latest/serde_json/enum.Value.html#impl-Display-for-Value) from the [Display](https://docs.rs/serde_json/latest/serde_json/enum.Value.html#impl-Display-for-Value) trait implementation.

With `Display` implemented, we can use the [write!](https://doc.rust-lang.org/std/macro.write.html) macro to write to any buffer type that implements some common traits. The `BytesMut` has implemented the necessary trait. So we can do something like the following to achieve the goal
```
let obj = Value::Object(...);
let mut buf = BytesMut::new();
write!(&mut buf, "{}", obj);

```